[Phoong Xiang Han] iP

xhphoong opened the pull request:
# **Anya**
> "Your mind is for having ideas, not holding them." - David Allen  
> [(source)](https://dansilvestre.com/productivity-quotes/)

Anya frees your mind of having to remember things you need to do. 
It's,
- Text-based
- *EASY* to learn
- ~FAST~ ***SUPER FAST*** to use

All you need to do is, 
1. Download it from [here](https://github.com/xhphoong/ip/releases/tag/v.02-A-Release)
2. Double-click it.
3. Add your tasks - **Todo**, **Deadline**, **Event**. User Guide can refer to [here](https://xhphoong.github.io/ip/)
4. Let Anya manage your tasks for you! 💯 👍 😄 

And it is **FREE**!

Features: 
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders

If you are a Java programmer, you can also use it to practice Java too. Here's the ``` main ``` method: 

```ruby 

 public static void main(String[] args) {
        Application.launch(Main.class, args);
 }

```

Commit:
Add Gradle support

Commit:
added Level 1

Commit:
added Level 2

Commit:
added Level 3

Commit:
added Level 4

Commit:
added Level 5

Commit:
added Level 6

Commit:
added A-TextUiTesting

Commit:
ignore(add pathway.txt)

Commit:
Revert "ignore(add pathway.txt)"

This reverts commit 576faf66b98404f45165759b8874af2da9b95c5d.

Commit:
added Level 7

Commit:
added Level 8

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8'

# Conflicts:
#	data/duke.txt
#	src/main/java/SkeletonDuke.java

Commit:
added A-MoreOOP

Commit:
added A-Packages

Commit:
Merge add-gradle-support branch

Commit:
added A-Gradle

Commit:
added A-JUnit

Commit:
ignore minor changes

Commit:
added A-JavaDoc

Commit:
added A-CodingStandard

Commit:
added Level 9

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Storage.java
#	src/main/java/Task.java
#	src/main/java/Ui.java

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	data/duke.txt
#	src/main/java/TaskList.java

Commit:
Add more OOP : Parser class, Command class , AddCommand class, DeleteCommand class, ExitCommand class, MarkCommand class, UnmarkCommand class, ListCommand class, FindCommand class

junwei-tan commented on src/main/java/AddCommand.java:
```diff
@@ -0,0 +1,27 @@
+ */
+public class AddCommand extends Command {
+
+    private final Task task;
```
Apt use of access modifiers.

junwei-tan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,62 @@
+    private TaskList tasks;
+    private Ui ui;
+    private Parser parser;
+    private final static String FILE_LOCATION = "./data/duke.txt";
```
Adherence to naming convention for constants.

junwei-tan commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,80 @@
+package duke;
```
Appreciate your use of packages.

junwei-tan reviewed (commented):
Overall a very readable application! I had a great time reviewing your PR.

Commit:
Add GUI to Duke

Commit:
Merge branch 'branch-Level-10'

doimoiboi commented on src/main/java/AddCommand.java:
```diff
@@ -0,0 +1,29 @@
+package duke;
```
Any reason you packaged all under a singular duke package instead of splitting the different classes into more packages? i.e. duke.commands, duke.entities

doimoiboi commented on src/main/java/MarkCommand.java:
```diff
@@ -0,0 +1,29 @@
+package duke;
+
+/**
```
Would it be better if the MarkCommand class could mark or unmark tasks? Thereby eliminating the need for an UnmarkTask class.

doimoiboi commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,77 @@
+            return new DeleteCommand(taskNo);
+        } else if (command.equals("find")) {
+            String keyword = strArr[1];
+            return new FindCommand(keyword);
```
Perhaps it would make for neater code to use an enum for the different command inputs?

doimoiboi commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+        }
+    }
+
+    /**
+     * Add new task from the users into the existing taskList.
+     * Keep track of the number of tasks in the list.
+     * Update the keyword hashtable.
+     * @param Task task: Task given by the users to add into the list.
+     * @return int noOfTasks: the current number of tasks in the list after adding it to the list.
+     */
```
Perhaps the comments would be easier to read with a blank line between the method description and the @params?

doimoiboi reviewed (commented):
LGTM! Overall really organised code and I just suggested some personal preferences.

Commit:
Add assertions to all Command's child classes, Storage class and TaskList class.

There is no existing assertion.

Assertions are used to define assumptions about the program state so that during the runtime, it can indicates a possible bug in the code.

Let's ,
*add assertions about the isExit boolean value at execute methods in Command's child classes. Doing so will ensure that the respective command's isExit value is correct
*add assertion about the strings that is read from a file at readFile() method in Storage class. Doing so will ensure that the line read is in correct format.
*add assertions about the noOfTasks value at add and delete method in TaskList class. Doing so will ensure that there is no negative noOfTasks.

xhphoong committed:
Merge pull request #1 from xhphoong/branch-A-Assertions

Add assertions to all Command's child classes, Storage class and TaskList class

Commit:
Improve code quality in Command and its child class by removing unused variable

*The Command and its child class contain isExit attribute that is not used at all.
*The Command's execute method pass in storage parameter that is not used in the method.

These unused parameter and attribute may cause confusion for the readers.

Let's,
*remove the isExit attributes in Command and all its child classes.
*remove the storage parameter in the execute method in the Command and all its child class.
*update the execute call in the Duke class by passing only task and ui.

Commit:
Improve Code Quality in by avoiding long methods, deep nesting, complicated expressions and code duplication.

*Deadline and Event class : toString method contains complicated expressions in the return statement.
*Parser : parse method is long and splitDescriptionAndDate method is deeply nested.
*Storage : readFile method is long.
*TaskList : findTaskWithThisKeyword contains a print line.
*Ui : printList and printMatchingList contains duplicated code.

Avoiding long method, deep nesting, complicated expression and code duplication can improve the readability of the code.

Let's,
*Extract out the time.format() in the return statement in the Deadline and Event's toString method. Evaluate it before the return statement.
*Extract out the code fragment related to addCommand in parse method in Parser class. Turn this code fragment into a new parseAddCommand method.
*Use Guard Clauses in splitDescriptionAndDate method in Parser class.
*Extract out the code fragment that process the String into task in the readFile method in Storage class. Turn this code fragment into a new readTask method.
*Remove the print statement in the findTaskWithThisKeyword method in TaskList class. Put the print statement in printMatchingList method in Ui class.
*Extract out duplicated code in the printList and printMatchingList method in Ui class. Create a new getPrintedList method that contains the duplicated code.

Commit:
Improve Code Quality by structuring code logically.

There are no blank lines in between the code.

Adding blank lines in between certain code can improve code's readability.

Let's add blank lines in between codes to group to related statements together.

xhphoong committed:
Merge pull request #2 from xhphoong/branch-A-CodeQuality

Improve CodeQuality

Commit:
Set up Continuous Integration (CI)

Commit:
Add CheckStyle

Use CheckStyle to detect code style violations and correct it accordingly.

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Add SortDeadline command

The deadlines is not sorted by date.

Adding a SortDeadline command can make the application more user-friendly because users now can call this command to see a list of deadlines that is sorted by dates. The deadline with the earliest deadline will be at the top of the list and the latest deadline will be at the bottom.

Let's,
* Add a TimeComparator class to compare the time.
* Add SortDeadlineCommand class, update Ui and Parser class.
* Add attribute : a list of deadlines in the taskList class and add the respective method to update and get deadlines list.

Commit:
Merge branch 'branch-C-Sort'

Commit:
Improve Gui

-Improve Gui to enhance the user experience.
-Make the conversation between user and chatbot asymmetric.
-Make the error message different to stand out to the users.
-Change padding, fonts, colors to improve appearance of GUI

Let's,
-Create 3 different DialogueBox, 1 for user, 1 for chatbot and 1 for showing errors.
-Change the padding for the dialog.
-Change background color, dialog color,textfield and send button color.
-Change font type and size.

Commit:
Set up a unique personality

Let's,
Change name from Duke to Anya.
Change some replies for Anya.

Commit:
Add more error handling

The error handled is mismatch input and task with no description. More errors can be handled.

Let's,
- Group all the exception into one exception - AnyaException.
- Handle extra error such as deadline/event with no dates, task number given is out of the taskList.
- Handle DateTimeParseException.

Commit:
Fix Junit Tests

Commit:
Fix Checkstyle

Commit:
Fix MissingDescription error

Commit:
Fix test error

Commit:
Fix gradle after changing personality of chatbot

Commit:
Add credits for this project

Commit:
Add user guide and ui.png

Commit:
Fix demo picture

Commit:
Add a help command

Commit:
Set up custom font

Commit:
Fix checkstyle issue

Commit:
Try to fix image not displaying issue

Commit:
Fix image pathway

Commit:
Add exception

Commit:
Modify the file location for storage and disallow resizing

Commit:
Add exception for read file

xhphoong committed:
Update README.md

xhphoong committed:
Update README.md

Commit:
Update credits

Commit:
Update credits.txt
